waveorder/util.py

```diff
@@ -871,52 +872,202 @@ def Dual_variable_ADMM_TV_deconv_2D(AHA, b_vec, rho, lambda_u, lambda_p, itr, ve
+        reg_x = xp.zeros(4)
+        reg_x[0] = reg_x_cent - search_range_auto  # search range = reg_x_cent +/- search_range_auto
+        reg_x[3] = reg_x_cent + search_range_auto
+        reg_x[1] = calc_golden_x(reg_x[0], reg_x[3])
+        reg_x[2] = reg_x[0] + (reg_x[3]-reg_x[1])
+
+        # holds the 4 current points
+        curr_pts = []
+        for i in range(4):
+            curr_pts.append(eval_L_curve(reg_x[i]))
+        
+        last_opt = None # only save the last point to save GPU memory
+        itr = 0
+        search_range = curr_pts[3].reg_x - curr_pts[0].reg_x
+        while search_range > epsilon_auto:
+            C1 = menger_curvature(curr_pts[:3])
+            C2 = menger_curvature(curr_pts[1:])
+            
+            # make sure right curvature is positive!
+                # keep moving the right edge inwards (and adjusting the middle points accordingly)
+            while C2 < 0:
+                new_reg_x = calc_golden_x(curr_pts[0].reg_x, curr_pts[2].reg_x)
+                curr_pts[3] = curr_pts[2]
+                curr_pts[2] = curr_pts[1]
+                curr_pts[1] = eval_L_curve(new_reg_x)
+                C2 = menger_curvature(curr_pts[1:])
+            C1 = menger_curvature(curr_pts[:3])
+
+            # case 1: left 3 points are better
+                # [a, b, c, d] --> [a, (a*phi+c)/(1+phi), b, c]
+            if C1 > C2:
+                last_opt = curr_pts[1]
+                new_reg_x = calc_golden_x(curr_pts[0].reg_x, curr_pts[2].reg_x)
+                curr_pts[3] = curr_pts[2]
+                curr_pts[2] = curr_pts[1]
+                curr_pts[1] = eval_L_curve(new_reg_x)
+            
+            # case 2: right 3 points are better
+                # [a, b, c, d] --> [b, c, b+d-c, d]
+            else:
+                last_opt = curr_pts[2]
+                new_reg_x = curr_pts[1].reg_x + curr_pts[3].reg_x - curr_pts[2].reg_x
+                curr_pts[0] = curr_pts[1]
+                curr_pts[1] = curr_pts[2]
+                curr_pts[2] = eval_L_curve(new_reg_x)
+            
+            itr += 1
+            search_range = curr_pts[3].reg_x - curr_pts[0].reg_x
+            if verbose:
+                print('Iteration: %d, deviation of the regularization interval: %.2e'%(itr, search_range))
         
-    return f_real
+        if verbose:
+            print('Final regularization parameter chosen: %.2e' % last_opt.reg)
+        
+            
+        # Return 3 solutions: the parameter chosen by the L-curve +/- epsilon
+            # try to not keep intermediate values in memory
+        f_real = []
+        f_real.append(ifft_f_real(compute_f_real_f(last_opt.reg_x - epsilon_auto)))
+        f_real.append(ifft_f_real(compute_f_real_f(last_opt.reg_x)))
+        f_real.append(ifft_f_real(compute_f_real_f(last_opt.reg_x + epsilon_auto)))
+
+        if output_lambda:
+            return np.array(f_real), last_opt.reg_x
```

Comment:
Hey @camFoltz ,

last_opt.reg_x is log_10 of the regularization parameter. If you wish to use this number for reg_re, you would need to turn it this way: 
`reg_re = 10**(last_opt.reg_x)`
Otherwise, this parameter pullout looks good to me.

Reply:
Sounds good, I just updated this to output the non-log scale.  Thanks!